### DISCUSSION TITLE
upgrade process?

### DISCUSSION THREAD

**Discussion by itsits** [jasonacox/Powerwall-Dashboard#17] [Q&A] [answered]
Is there a process to upgrade from an earlier version without losing data?

**Comment by jasonacox** [comment, marked as answer]
That's a great question @itsits  - The data is stored in the influxdb folder.  I should add an `upgrade.sh` to the root folder.  It would basically be this:

```bash
#!/bin/bash

echo "Upgrade Powerwall-Dashboard"
echo "---------------------------"
echo "This script will upgrade you to the latest version without"
echo "removing existing data. A backup is still recommended."
echo ""

read -r -p "Upgrade - Proceed? [y/N] " response
if [[ "$response" =~ ^([yY][eE][sS]|[yY])$ ]]
then
    echo ""
else
    echo "Cancel"
    exit 
fi

echo "Pull influxdb.sql, dashboard.json, telegraf.conf, and other changes..."
git pull 

echo ""
echo "Add downsample continuous queries..."
docker exec -it influxdb influx -import -path=/var/lib/influxdb/influxdb.sql

echo ""
echo "Delete and Upgrade pyPowerwall to Latest"
docker stop pypowerwall
docker rm pypowerwall
docker images | grep pypowerwall | awk '{print $3}' | xargs docker rmi -f

echo "Restarting Powerwall-Dashboard stack..."
docker-compose -f powerwall.yml up -d

# Display Final Instructions
cat << EOF
------------------[ Update Dashboard ]-----------------
Open Grafana at http://localhost:9000/ 

* From 'Dashboard\Manage' (or 'Dashboard\Browse'), select 'Import', and upload 'dashboard.json' from
EOF
pwd

echo ""
echo "Done"

```

**Reply by jasonacox** [reply]
Added [upgrade.sh](https://github.com/jasonacox/Powerwall-Dashboard/blob/main/upgrade.sh) script to root folder.

**Reply by nhasan** [reply]
I just tried this and git pull aborts as my  powerwall.yml is locally modified to specify my powerwall config. Any suggestions?
```
Updating ede926e..bbdb2ef
error: Your local changes to the following files would be overwritten by merge:
        powerwall.yml
Please commit your changes or stash them before you merge.
Aborting
```
Also, the dashboard has Los Angeles timezone. So the upgrade is not so seamless it seems.

**Reply by jasonacox** [reply]
That's a great point.  Thanks @nhasan - I suppose you could do:

```bash
git stash
git pull
git stash pop
```

But that still wouldn't work 100% because of the Timezone issue.  I need to come up with a better upgrade path.  Any suggestions or PRs welcome. ;)

**Reply by nhasan** [reply]
In powerwall.yml, you could use 'env_file' option to externalize the environment variables. Then in setup.sh you could create this environment file. If the file already exists, then this step could be skipped.

For dashboard.json, you could ask for the timezone and influxdb database name in setup.sh and do the substitution.

Of course this implies that one has to rerun setup.sh after every upgrade.

Edit: It seems the tz.sh already exists that does timezone substitution. However it has the same sed command line issue I reported earlier.

**Reply by nhasan** [reply]
I have sent a PR your way this morning. Hope you like it.

**Reply by jasonacox** [reply]
Thank you @nhasan , very elegant! 

I made two minor updates:  1) Added prompt for user to update the credentials during setup and 2) added a shell pypowerwall.env file for easier manual setup.

**Reply by jasonacox** [reply]
@nhasan I updated pypowerwall proxy to `t8` to include the "backup switch" update - it would be interesting to see if upgrade.sh now works for you.

**Reply by nhasan** [reply]
@jasonacox  Thanks for considering my changes. I have one issue though, the pypowerwall.env file should not be checked in. The entire premise of this change depends on that file not being in git. Please see my comment on the commit for explanation.

**Reply by jasonacox** [reply]
Closing this out.  You are correct @nhasan, making `pypowerwall.env` external is the most elegant path.  Thanks for the help and PRs to get this set up.  Setup script `setup.sh` and README instructions have been updated to reflect the changes.  Using `upgrade.sh` should work for most users now.

ALL: Please open an issue or provide feedback here if you encounter any issues.
